### PR TITLE
Make nonlinear solver intercept model assignments from the linear arithmetic solver

### PR DESCRIPTION
**Pull request by ajreynol**
This fixes consistency issues for models generated for non-linear arithmetic. 

Currently, the models generated by nonlinear arithmetic have "internal inconsistencies", e.g. the equality engine of the TheoryModel object contains assertions that it does not satisfy. This is caused by the following chain of events:

(1) The linear arithmetic solver decides on a set of model assignments and adds them to the current candidate model object, call it M. In the implementaiton, this is communicated in the form of a set of equalities v1 = c1 ^ ... ^ vn = cn which are added to the TheoryModel object during collectModelInfo.
(2) During a last call effort, the nonlinear solver may infer that the current assertions are in fact satisfied by an alternative model assignment, that is, for some different v1 = c'1 ^ ... ^ vn = c'n.  It then "overwrites" these values by the recordApproximation feature.  
(3) The system (soundly) answers "SAT". The model is in fact guaranteed to be correct wrt getModelValue calls. However, the model is "internally inconsistent" since the original equalities v1 = c1 ^ ... ^ vn = cn are not retracted from the model. Thus, we encounter assertion failures in debug as well as `--check-model` failures in either production and debug (these check model failures only occur for internally generated assertions).

This commit refactors the interplay between the linear and nonlinear solvers so that the nonlinear solver "intercepts" the model assignment from the linear solver and modifies it as necessary *before* it is sent to TheoryModel during collectModelInfo.  This ensures that no assertions are later invalidated when the nonlinear solver decides to change model values based on nonlinear reasoning a la Cimatti et al TACAS 2017.

This PR also removes the use of "skolem atoms" which was a way of avoiding checkModel failures due to internally introduced atoms. This is no longer necessary as we can correctly generate models now.

This fixes #2531 (that benchmark times out now).  We now generate internally consistent models for the benchmarks on #2008, #3003, #3441, #3411, which are added as regressions by this commit.
Notice that `--check-model` fails on some benchmarks still, since we do not have the proper infrastructure in place for aborting checkModel calls when approximations are used. I've reopened #2008 to track this issue.

### CODE REVIEW

**Comment by 4tXJ7f** [comment]
I am adding @alex-ozdemir as a secondary reviewer since he is interested in starting to review PRs.

**Comment by ahmed-irfan** [comment]
Shouldn't there be some regression tests with the option `--check-model`?

**Comment by ajreynol** [comment]
`--check-model` is run for all regression tests by default :)
See https://github.com/CVC4/CVC4/blob/master/test/regress/run_regression.py#L309

File: test/regress/regress1/nl/issue3441.smt2 [hunk 1]
```diff
@@ -0,0 +1,8 @@
+(set-logic QF_NIA)
```
Comment (ahmed-irfan) [inline test/regress/regress1/nl/issue3441.smt2]: I am not sure, but did you miss the headers 
```
; COMMAND-LINE
; EXPECT
```

Comment (ajreynol) [reply test/regress/regress1/nl/issue3441.smt2]: This isn't necessary if there aren't any specific options you are using or output you are expecting (beyond a status).

**Comment by ahmed-irfan** [comment]
> `--check-model` is run for all regression tests by default :)
> See https://github.com/CVC4/CVC4/blob/master/test/regress/run_regression.py#L309

I didn't know :)
Thanks

**Pull request merged**